File: second_moment/membership.py
"""membership.py -- the exact membership test on (|S|, sigma_t(S), sum x_u^2) against the (Q-CERT+) threshold

Supports, in "Certificates for short extending words in a finite automaton":
  Appendix D: "On the exhaustive synchronizing strongly connected binary
    populations, one row per proper nonempty subset and length t <= n-1, it
    certifies 0 rows of 708 at n=3 that the threshold misses, 80 of 52,080 at
    n=4, 15,053 of 3,910,560 at n=5 and 2,083,316 of 347,983,990 at n=6, that
    is 0, 0.15, 0.38 and 0.60 percent, rising at each size" -- this program
    computes n=3, 4, 5; membership.rs computes n=6 (and n<=5 as a cross-check).
  Appendix D: "at k^t=4, |S|=3 and -sigma_t(S)=3 the values attainable with
    every x_u<=0 are 3, 5 and 9", with 9 - 5 = 4 = 2(|S|-1).
Population / convention: binary automata on Q = {0,...,n-1}; letter a over one
  representative of each conjugacy class of endofunctions of Q (7, 19, 47
  classes at n=3, 4, 5), letter b over all n^n maps; kept if synchronizing and
  strongly connected (59, 1,240 and 32,588 automata: the quotient convention of
  Appendix A.1).  One row per kept automaton, proper nonempty subset S and
  length t = 1..n-1, so rows = automata x (2^n - 2) x (n-1).
  For a row: x_u = |S u^{-1}| - |S| over the 2^t words u of length t (applied
  letter by letter), m = -sigma_t(S) = -sum x_u, c = |S|, q = floor(m/c),
  r = m - qc.
    THRESHOLD (Q-CERT+, Proposition 4) fires iff sum x_u^2 > q c^2 + r^2
              (always when m < 0).
    MEMBERSHIP fires iff sum x_u^2 is not a value of sum y_u^2 over integers
              y_u in [0, c] (one per word) with sum y_u = m, i.e. iff the data
              (c, sigma_t(S), sum x_u^2) cannot come from x_u <= 0 for every u.
  A row counted in memonly_rows_syncsc is one where MEMBERSHIP fires and the
  THRESHOLD does not.
Usage:    python membership.py N          (N = 3, 4 or 5)
Output:   KEY=VALUE counters, with the keys printed by membership.rs: classes,
          autos (automata enumerated), syncsc (kept), eul (Eulerian among them),
          rows, thr, mem, memonly, memonly_sync (rows of kept automata only),
          rows_syncsc, thr_syncsc, memonly_rows_syncsc; then one [PASS]/[FAIL]
          row per checked statement and "R rows, F FAILED"; exit status F.
          Appendix D's figures are memonly_rows_syncsc of rows_syncsc.
Runtime:  one core: n=3 and n=4 about 1 s, n=5 about 30 s.
Requires: Python 3 (standard library only).
"""
import itertools
import sys
import time
from functools import lru_cache


# ---------- endofunction conjugacy classes under S_n
def canon(f, n, perms):
    best = None
    for p in perms:
        inv = [0] * n
        for i, pi in enumerate(p):
            inv[pi] = i
        g = tuple(p[f[inv[i]]] for i in range(n))
        if best is None or g < best:
            best = g
    return best


def classes(n):
    perms = list(itertools.permutations(range(n)))
    seen, reps = set(), []
    for f in itertools.product(range(n), repeat=n):
        c = canon(list(f), n, perms)
        if c not in seen:
            seen.add(c)
            reps.append(list(f))
    return reps


# ---------- automaton predicates
def is_sync(a, b, n):
    """pair-automaton reachability: synchronizing iff every pair merges"""
    frontier = {(p, q) for p in range(n) for q in range(p + 1, n)}
    merged = set()
    for pair in frontier:
        seen = {pair}
        stack = [pair]
        ok = False
        while stack:
            p, q = stack.pop()
            if p == q:
                ok = True
                break
            for f in (a, b):
                nx = (min(f[p], f[q]), max(f[p], f[q]))
                if nx not in seen:
                    seen.add(nx)
                    stack.append(nx)
        if not ok:
            return False
    return True


def is_sc(a, b, n):
    fwd = [set() for _ in range(n)]
    rev = [set() for _ in range(n)]
    for q in range(n):
        for f in (a, b):
            fwd[q].add(f[q])
            rev[f[q]].add(q)

    def reach(g, s):
        seen, st = {s}, [s]
        while st:
            x = st.pop()
            for y in g[x]:
                if y not in seen:
                    seen.add(y)
                    st.append(y)
        return seen
    return len(reach(fwd, 0)) == n and len(reach(rev, 0)) == n


# ---------- attainable sum-of-squares under all-non-positivity
@lru_cache(maxsize=None)
def attainable(N, c, m):
    """set of sum(y^2) over y in [0,c]^N with sum(y)=m"""
    if m < 0 or m > N * c:
        return frozenset()
    cur = {0: {0}}
    for _ in range(N):
        nxt = {}
        for s, sq in cur.items():
            for y in range(0, c + 1):
                s2 = s + y
                if s2 > m:
                    break
                d = nxt.setdefault(s2, set())
                yy = y * y
                for v in sq:
                    d.add(v + yy)
        cur = nxt
    return frozenset(cur.get(m, ()))


def run(n, tmax, want_filter):
    """returns counters over the quotient-convention corpus"""
    reps = classes(n)
    allmaps = list(itertools.product(range(n), repeat=n))
    subsets = [tuple(s) for size in range(1, n)
               for s in itertools.combinations(range(n), size)]
    C = dict(autos=0, sync=0, sc=0, syncsc=0, eul=0, rows=0,
             thr=0, mem=0, memonly=0, memonly_sync=0,
             memonly_rows_syncsc=0, rows_syncsc=0, thr_syncsc=0,
             bneg_rows=0, memonly_bneg=0)
    for a in reps:
        for bt in allmaps:
            b = list(bt)
            C["autos"] += 1
            sy = is_sync(a, b, n)
            sc = is_sc(a, b, n)
            if sy:
                C["sync"] += 1
            if sc:
                C["sc"] += 1
            keep = sy and sc
            if keep:
                C["syncsc"] += 1
                indeg = [0] * n
                for q in range(n):
                    for f in (a, b):
                        indeg[f[q]] += 1
                if all(d == 2 for d in indeg):
                    C["eul"] += 1
            if want_filter and not keep:
                continue
            # per-word preimage-size vectors, per length t
            for t in range(1, tmax + 1):
                words = list(itertools.product((0, 1), repeat=t))
                vs = []
                for w in words:
                    v = [0] * n
                    for q in range(n):
                        x = q
                        for ch in w:
                            x = (a if ch == 0 else b)[x]
                        v[x] += 1
                    vs.append(v)
                for S in subsets:
                    c = len(S)
                    xs = [sum(v[r] for r in S) - c for v in vs]
                    sig = sum(xs)
                    sq = sum(x * x for x in xs)
                    m = -sig
                    C["rows"] += 1
                    if keep:
                        C["rows_syncsc"] += 1
                    # threshold (Q-CERT+)
                    if m < 0:
                        thr = True
                    else:
                        q_, r_ = divmod(m, c)
                        thr = sq > q_ * c * c + r_ * r_
                    mem = sq not in attainable(len(words), c, m)
                    if thr:
                        C["thr"] += 1
                        if keep:
                            C["thr_syncsc"] += 1
                    if mem:
                        C["mem"] += 1
                    if mem and not thr:
                        C["memonly"] += 1
                        if sy:
                            C["memonly_sync"] += 1
                        if keep:
                            C["memonly_rows_syncsc"] += 1
    return C


# the paper's values: n -> (automata, Eulerian, rows, membership-only rows, share as printed)
PAPER = {
    3: (59, 8, 708, 0, "0"),
    4: (1240, 72, 52080, 80, "0.15"),
    5: (32588, 954, 3910560, 15053, "0.38"),
}
KEYS = ("classes", "autos", "syncsc", "eul", "rows", "thr", "mem", "memonly",
        "memonly_sync", "rows_syncsc", "thr_syncsc", "memonly_rows_syncsc")


def main(argv):
    if len(argv) != 2 or argv[1] not in ("3", "4", "5"):
        sys.exit("usage: python membership.py N    (N = 3, 4 or 5; n = 6: membership.rs)")
    n = int(argv[1])
    t0 = time.time()
    C = run(n, n - 1, want_filter=True)
    C["classes"] = len(classes(n))
    wall = time.time() - t0
    print("n=%d" % n)
    for key in KEYS:
        print("%s=%d" % (key, C[key]))
    print("wall=%.1fs" % wall)
    print("")

    checks = []

    def row(ok, text):
        checks.append(ok)
        print("[%s] %s" % ("PASS" if ok else "FAIL", text))

    autos, eul, rows, memonly, share = PAPER[n]
    got_share = 100.0 * C["memonly_rows_syncsc"] / C["rows_syncsc"] if C["rows_syncsc"] else float("nan")
    got_share_txt = "0" if C["memonly_rows_syncsc"] == 0 else "%.2f" % got_share
    row(C["syncsc"] == autos and C["eul"] == eul,
        "Appendix A, n=%d binary synchronizing strongly connected population: %s automata, "
        "%s Eulerian (paper: %s, %s)" % (n, "{:,}".format(C["syncsc"]), "{:,}".format(C["eul"]),
                                          "{:,}".format(autos), "{:,}".format(eul)))
    row(C["rows_syncsc"] == rows == autos * (2 ** n - 2) * (n - 1),
        "Appendix D, n=%d: %s subset-length rows = automata x (2^n-2) x (n-1) (paper: %s)"
        % (n, "{:,}".format(C["rows_syncsc"]), "{:,}".format(rows)))
    row(C["memonly_rows_syncsc"] == memonly,
        "Appendix D, n=%d: membership certifies %s rows the (Q-CERT+) threshold misses (paper: %s of %s)"
        % (n, "{:,}".format(C["memonly_rows_syncsc"]), "{:,}".format(memonly), "{:,}".format(rows)))
    row(got_share_txt == share,
        "Appendix D, n=%d: share %.4f percent, printed %s (paper: %s percent)"
        % (n, got_share, got_share_txt, share))
    vals = sorted(attainable(4, 3, 3))
    row(vals == [3, 5, 9],
        "Appendix D: k^t=4, |S|=3, -sigma_t(S)=3: second moments attainable with every x_u<=0 = %s "
        "(paper: 3, 5 and 9)" % vals)
    q_, r_ = divmod(3, 3)
    thr = q_ * 3 * 3 + r_ * r_
    below = max(v for v in vals if v < thr)
    row(thr == 9 and 7 not in vals and thr - below == 4 == 2 * (3 - 1),
        "Appendix D: there the (Q-CERT+) threshold is %d, the observation 7 is not attainable, and the "
        "largest attainable value below the threshold falls short by %d-%d = %d = 2(|S|-1) "
        "(paper: 9-5=4=2(|S|-1))" % (thr, thr, below, thr - below))
    failed = checks.count(False)
    print("%d rows, %d FAILED" % (len(checks), failed))
    return min(failed, 255)


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: second_moment/membership.rs
// membership.rs -- the exact membership test on (|S|, sigma_t(S), sum x_u^2)
// against the (Q-CERT+) threshold, binary automata, any n (used at n = 6)
//
// Supports, in "Certificates for short extending words in a finite automaton":
//   Appendix D: the membership test "certifies ... 2,083,316 of 347,983,990 at
//     n=6, that is ... 0.60 percent" (rows of the exhaustive synchronizing
//     strongly connected binary population, one per proper nonempty subset and
//     length t <= n-1, on which the membership test fires and the (Q-CERT+)
//     threshold does not).  The same computation as membership.py, which covers
//     n = 3, 4, 5 (0 of 708, 80 of 52,080, 15,053 of 3,910,560); at those sizes
//     the two programs print identical counters.
// Population / convention: letter a over one representative of each conjugacy
//   class of endofunctions of Q = {0,...,n-1} (130 classes at n=6), letter b
//   over all n^n maps; with "filter" only synchronizing strongly connected
//   automata get rows (1,122,529 at n=6: quotient convention, Appendix A.1).
//   Row (S, t): x_u = |S u^{-1}| - |S| over the 2^t words of length t,
//   m = -sigma_t(S), c = |S|.  The threshold fires iff sum x_u^2 > q c^2 + r^2
//   with q = floor(m/c), r = m - qc (always when m < 0); membership fires iff
//   sum x_u^2 is not attainable as sum y_u^2 with 2^t integers y_u in [0, c]
//   summing to m.
// Usage:    rustc -O membership.rs            (produces ./membership)
//           ./membership 6 filter             ("nofilter": rows for every automaton)
// Output:   one KEY=VALUE per line: n, classes, autos (enumerated), syncsc,
//           eul, rows, thr, mem, memonly, memonly_sync, rows_syncsc, thr_syncsc,
//           memonly_rows_syncsc.  Appendix D's n=6 figures are
//           memonly_rows_syncsc=2083316 and rows_syncsc=347983990.
// Runtime:  one core: n=6 filter 2 to 3 minutes (167 s measured on a shared
//           machine); n <= 5 under 2 s.
// Requires: rustc (tested with 1.97), standard library only.

use std::collections::HashMap;
use std::env;

fn perms(n: usize) -> Vec<Vec<usize>> {
    let mut out = Vec::new();
    let mut p: Vec<usize> = (0..n).collect();
    loop {
        out.push(p.clone());
        // next permutation
        let mut i = n as isize - 2;
        while i >= 0 && p[i as usize] >= p[i as usize + 1] {
            i -= 1;
        }
        if i < 0 {
            break;
        }
        let i = i as usize;
        let mut j = n - 1;
        while p[j] <= p[i] {
            j -= 1;
        }
        p.swap(i, j);
        p[i + 1..].reverse();
    }
    out
}

/// canonical form of an endofunction under simultaneous conjugation
fn canon(f: &[usize], n: usize, ps: &[Vec<usize>]) -> Vec<usize> {
    let mut best: Option<Vec<usize>> = None;
    let mut inv = vec![0usize; n];
    for p in ps {
        for i in 0..n {
            inv[p[i]] = i;
        }
        let g: Vec<usize> = (0..n).map(|i| p[f[inv[i]]]).collect();
        if best.is_none() || g < *best.as_ref().unwrap() {
            best = Some(g);
        }
    }
    best.unwrap()
}

fn classes(n: usize) -> Vec<Vec<usize>> {
    let ps = perms(n);
    let total = (n as u64).pow(n as u32);
    let mut seen = std::collections::HashSet::new();
    let mut reps = Vec::new();
    for code in 0..total {
        let mut f = vec![0usize; n];
        let mut c = code;
        for i in 0..n {
            f[i] = (c % n as u64) as usize;
            c /= n as u64;
        }
        let k = canon(&f, n, &ps);
        if seen.insert(k) {
            reps.push(f);
        }
    }
    reps
}

fn is_sync(a: &[usize], b: &[usize], n: usize) -> bool {
    // every unordered pair must reach the diagonal
    for p0 in 0..n {
        for q0 in (p0 + 1)..n {
            let mut seen = vec![false; n * n];
            let mut stack = vec![(p0, q0)];
            seen[p0 * n + q0] = true;
            let mut ok = false;
            while let Some((p, q)) = stack.pop() {
                if p == q {
                    ok = true;
                    break;
                }
                for f in [a, b] {
                    let (mut x, mut y) = (f[p], f[q]);
                    if x > y {
                        std::mem::swap(&mut x, &mut y);
                    }
                    if !seen[x * n + y] {
                        seen[x * n + y] = true;
                        stack.push((x, y));
                    }
                }
            }
            if !ok {
                return false;
            }
        }
    }
    true
}

fn is_sc(a: &[usize], b: &[usize], n: usize) -> bool {
    let mut fwd = vec![Vec::new(); n];
    let mut rev = vec![Vec::new(); n];
    for q in 0..n {
        for f in [a, b] {
            fwd[q].push(f[q]);
            rev[f[q]].push(q);
        }
    }
    let reach = |g: &Vec<Vec<usize>>| -> usize {
        let mut seen = vec![false; n];
        let mut st = vec![0usize];
        seen[0] = true;
        let mut c = 1;
        while let Some(x) = st.pop() {
            for &y in &g[x] {
                if !seen[y] {
                    seen[y] = true;
                    c += 1;
                    st.push(y);
                }
            }
        }
        c
    };
    reach(&fwd) == n && reach(&rev) == n
}

/// table[m][sq] = true iff sum(y^2)=sq is attainable with N terms in [0,c]
/// summing to m.  Built once per (N,c).
fn attain_table(nw: usize, c: usize) -> Vec<Vec<bool>> {
    let maxm = nw * c;
    let maxsq = nw * c * c;
    let mut cur = vec![vec![false; maxsq + 1]; maxm + 1];
    cur[0][0] = true;
    for _ in 0..nw {
        let mut nxt = vec![vec![false; maxsq + 1]; maxm + 1];
        for s in 0..=maxm {
            for sq in 0..=maxsq {
                if !cur[s][sq] {
                    continue;
                }
                for y in 0..=c {
                    let s2 = s + y;
                    let q2 = sq + y * y;
                    if s2 > maxm || q2 > maxsq {
                        break;
                    }
                    nxt[s2][q2] = true;
                }
            }
        }
        cur = nxt;
    }
    cur
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: usize = args[1].parse().unwrap();
    let want_filter = args.get(2).map(|s| s == "filter").unwrap_or(true);
    let tmax = n - 1;

    let reps = classes(n);
    let nsub = 1usize << n;

    // proper nonempty subsets as bitmasks
    let subsets: Vec<usize> = (1..nsub - 1).collect();

    let mut cache: HashMap<(usize, usize), Vec<Vec<bool>>> = HashMap::new();
    for t in 1..=tmax {
        let nw = 1usize << t;
        for c in 1..n {
            cache.entry((nw, c)).or_insert_with(|| attain_table(nw, c));
        }
    }

    let mut autos: u64 = 0;
    let mut syncsc: u64 = 0;
    let mut eul: u64 = 0;
    let mut rows: u64 = 0;
    let mut thr: u64 = 0;
    let mut mem: u64 = 0;
    let mut memonly: u64 = 0;
    let mut memonly_sync: u64 = 0;
    let mut rows_syncsc: u64 = 0;
    let mut thr_syncsc: u64 = 0;
    let mut memonly_syncsc: u64 = 0;

    let total_b = (n as u64).pow(n as u32);
    let mut b = vec![0usize; n];
    for a in &reps {
        for code in 0..total_b {
            let mut c = code;
            for i in 0..n {
                b[i] = (c % n as u64) as usize;
                c /= n as u64;
            }
            autos += 1;
            let sy = is_sync(a, &b, n);
            let sc = is_sc(a, &b, n);
            let keep = sy && sc;
            if keep {
                syncsc += 1;
                let mut indeg = vec![0usize; n];
                for q in 0..n {
                    indeg[a[q]] += 1;
                    indeg[b[q]] += 1;
                }
                if indeg.iter().all(|&d| d == 2) {
                    eul += 1;
                }
            }
            if want_filter && !keep {
                continue;
            }
            for t in 1..=tmax {
                let nw = 1usize << t;
                // per-word preimage-count vectors, then subset sums over masks
                // sums[w][mask] = |S u^{-1}| for the w-th word of length t
                let mut sums: Vec<Vec<i64>> = Vec::with_capacity(nw);
                for w in 0..nw {
                    let mut v = vec![0i64; n];
                    for q in 0..n {
                        let mut x = q;
                        let mut ww = w;
                        for _ in 0..t {
                            x = if ww & 1 == 0 { a[x] } else { b[x] };
                            ww >>= 1;
                        }
                        v[x] += 1;
                    }
                    // subset-sum over masks
                    let mut s = vec![0i64; nsub];
                    for mask in 1..nsub {
                        let low = mask.trailing_zeros() as usize;
                        s[mask] = s[mask & (mask - 1)] + v[low];
                    }
                    sums.push(s);
                }
                for &mask in &subsets {
                    let cs = mask.count_ones() as i64;
                    let mut sig: i64 = 0;
                    let mut sq: i64 = 0;
                    for w in 0..nw {
                        let x = sums[w][mask] - cs;
                        sig += x;
                        sq += x * x;
                    }
                    rows += 1;
                    if keep {
                        rows_syncsc += 1;
                    }
                    let m = -sig;
                    let fires_thr = if m < 0 {
                        true
                    } else {
                        let q_ = m / cs;
                        let r_ = m - q_ * cs;
                        sq > q_ * cs * cs + r_ * r_
                    };
                    let tab = &cache[&(nw, cs as usize)];
                    let fires_mem = if m < 0 || (m as usize) >= tab.len() {
                        true
                    } else {
                        let row = &tab[m as usize];
                        let s = sq as usize;
                        !(s < row.len() && row[s])
                    };
                    if fires_thr {
                        thr += 1;
                        if keep {
                            thr_syncsc += 1;
                        }
                    }
                    if fires_mem {
                        mem += 1;
                    }
                    if fires_mem && !fires_thr {
                        memonly += 1;
                        if sy {
                            memonly_sync += 1;
                        }
                        if keep {
                            memonly_syncsc += 1;
                        }
                    }
                }
            }
        }
    }

    println!("n={}", n);
    println!("classes={}", reps.len());
    println!("autos={}", autos);
    println!("syncsc={}", syncsc);
    println!("eul={}", eul);
    println!("rows={}", rows);
    println!("thr={}", thr);
    println!("mem={}", mem);
    println!("memonly={}", memonly);
    println!("memonly_sync={}", memonly_sync);
    println!("rows_syncsc={}", rows_syncsc);
    println!("thr_syncsc={}", thr_syncsc);
    println!("memonly_rows_syncsc={}", memonly_syncsc);
}
